tezos/new_context/src/working_tree/serializer.rs

```diff
@@ -204,6 +225,101 @@ pub fn serialize_entry(
     Ok(())
 }
 
+fn serialize_hash_id(hash_id: u32, output: &mut Vec<u8>) -> Result<usize, SerializationError> {
+    if hash_id & 0x7FFFFF == hash_id {
+        // The HashId fits in 23 bits
+
+        // Set `COMPACT_HASH_ID_BIT` so the deserializer knows the `HashId` is in 3 bytes
+        let hash_id: u32 = hash_id | COMPACT_HASH_ID_BIT;
+        let hash_id: [u8; 4] = hash_id.to_be_bytes();
+
+        output.write_all(&hash_id[1..])?;
+        Ok(3)
+    } else if hash_id & 0x7FFFFFFF == hash_id {
+        // HashId fits in 31 bits
+
+        output.write_all(&hash_id.to_be_bytes())?;
+        Ok(4)
+    } else {
+        // The HashId must not be 32 bits because we use the
+        // MSB to determine if the HashId is compact or not
+        Err(SerializationError::HashIdTooBig)
+    }
+}
+
+fn serialize_inode(
+    inode_id: InodeId,
+    output: &mut Vec<u8>,
+    hash_id: HashId,
+    storage: &Storage,
+    stats: &mut SerializeStats,
+    batch: &mut Vec<(HashId, Arc<[u8]>)>,
+    referenced_older_entries: &mut Vec<HashId>,
+) -> Result<(), SerializationError> {
+    output.clear();
+    let inode = storage.get_inode(inode_id).unwrap();
+
+    match inode {
+        Inode::Pointers {
+            depth,
+            nchildren,
+            npointers,
+            pointers,
+        } => {
+            output.write_all(&[ID_INODE_POINTERS])?;
+            output.write_all(&depth.to_ne_bytes())?;
+            output.write_all(&nchildren.to_ne_bytes())?;
+            output.write_all(&npointers.to_ne_bytes())?;
+
+            // TODO: Make a bitfield of 32 bits indicating how many hashes there are
+            //       and at what index, instead of wasting 1 byte (the index below) for
+            //       each hash
+            for (index, pointer) in pointers.iter().enumerate() {
+                if let Some(pointer) = pointer {
+                    let index: u8 = index as u8;
+
+                    let hash_id = pointer.hash_id();
+                    let hash_id = hash_id.map(|h| h.as_u32()).unwrap_or(0);
+
+                    output.write_all(&index.to_ne_bytes())?;
+                    serialize_hash_id(hash_id, output)?;
+                };
+            }
+
+            batch.push((hash_id, Arc::from(output.as_slice())));
+
+            for pointer in pointers.iter().filter_map(|p| p.as_ref()) {
+                let hash_id = pointer.hash_id().ok_or(SerializationError::MissingHashId)?;
+
+                if pointer.is_commited() {
```

Comment:
Explain here that we only want to serialize new entries, while also recognize al references to old entries because of the GC.

Reply:
I mean, with a comment in the code here